File: src/cli/commands/approval.ts
/** `jackops approval` command handler. */

import { join } from "@std/path";
import { APPROVAL_MODES, isApprovalMode } from "../../config.ts";
import * as worktree from "../../worktree.ts";
import * as daemon from "../../daemon.ts";
import { loadCliConfig } from "../context.ts";

export async function approval(args: string[]): Promise<void> {
  const [mode] = args;
  const base = Deno.cwd();
  const { config } = await loadCliConfig();

  if (!mode) {
    // Show current mode
    const runtime = await daemon.readApprovalMode(base);
    const current = runtime ?? config.orchestrator.approval;
    const source = runtime ? "runtime" : "config";
    console.log(`Approval mode: ${current} (${source})`);
    return;
  }

  if (!isApprovalMode(mode)) {
    console.error(
      `Invalid approval mode '${mode}'. Must be one of: ${
        APPROVAL_MODES.join(", ")
      }`,
    );
    Deno.exit(1);
  }

  // Rewrite Claude settings for each Claude worker
  const switched: string[] = [];
  const skipped: string[] = [];

  for (const w of config.workers) {
    if (w.agent === "claude") {
      const wt = join(
        base,
        worktree.worktreeDir(config.project, w.name),
      );
      await daemon.writeClaudeSettings(wt, base, w.name, mode);
      switched.push(w.name);
    } else {
      skipped.push(`${w.name} (${w.agent})`);
    }
  }

  // Update orchestrator settings if Claude
  if (config.orchestrator.agent === "claude") {
    await daemon.mergeClaudeSettings(base, base, "orchestrator", mode);
    switched.push("orchestrator");
  }

  // Write runtime mode file for daemon
  await daemon.writeApprovalMode(base, mode);

  console.log(`Approval mode switched to: ${mode}`);
  if (switched.length > 0) {
    console.log(`  Updated: ${switched.join(", ")}`);
  }
  if (skipped.length > 0) {
    console.log(
      `  Skipped (no live switching): ${skipped.join(", ")}`,
    );
  }
}

File: src/cmd/approval.go
package cmd

import (
	"fmt"

	"github.com/AdriaanVE/jack-in/src/daemon"
	"github.com/AdriaanVE/jack-in/src/domain"
	"github.com/AdriaanVE/jack-in/src/worktree"
	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(approvalCmd)
}

var approvalCmd = &cobra.Command{
	Use:   "approval [mode]",
	Short: "Show or switch approval mode (manual, auto, yolo)",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runApproval,
}

func runApproval(cmd *cobra.Command, args []string) error {
	base := cwd()
	cfg, err := loadConfig(base)
	if err != nil {
		return err
	}

	if len(args) == 0 {
		mode, source := resolveApproval(base, cfg)
		fmt.Printf("Approval mode: %s (%s)\n", mode, source)
		return nil
	}

	// Switch mode
	mode := domain.ApprovalMode(args[0])
	if !domain.IsApprovalMode(string(mode)) {
		return fmt.Errorf("invalid mode %q: must be manual, auto, or yolo", args[0])
	}

	var switched []string
	var skipped []string

	for _, w := range cfg.Workers {
		if w.Agent != domain.AgentClaude {
			skipped = append(skipped, w.Name)
			continue
		}
		wt := worktree.Path(base, cfg.Project, w.Name)
		if err := daemon.WriteClaudeSettings(wt, base, w.Name, mode, nil); err != nil {
			return fmt.Errorf("updating settings for %s: %w", w.Name, err)
		}
		switched = append(switched, w.Name)
	}

	// Update orchestrator settings if it uses Claude
	if cfg.Orchestrator.Agent == string(domain.AgentClaude) {
		if err := daemon.MergeClaudeSettings(base, base, "orchestrator", mode, daemon.OrchestratorPermissions); err != nil {
			return fmt.Errorf("updating orchestrator settings: %w", err)
		}
	}

	// Write runtime mode file
	if err := daemon.WriteApprovalMode(base, mode); err != nil {
		return fmt.Errorf("writing approval mode: %w", err)
	}

	if len(switched) > 0 {
		fmt.Printf("Switched to %s: %v\n", mode, switched)
	}
	if len(skipped) > 0 {
		fmt.Printf("Skipped (non-Claude): %v\n", skipped)
	}
	return nil
}
